src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProject.cs

```diff
@@ -0,0 +1,1146 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.VisualStudio.LanguageServices.Implementation.TaskList;
+using Roslyn.Utilities;
+
+namespace Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem
+{
+    internal sealed class VisualStudioProject
+    {
+        private readonly VisualStudioWorkspaceImpl _workspace;
+        private readonly HostDiagnosticUpdateSource _hostDiagnosticUpdateSource;
+        private readonly string _projectUniqueName;
+
+        /// <summary>
+        /// A gate taken for all mutation of any mutable field in this type.
+        /// </summary>
+        /// <remarks>This is, for now, intentionally pessimistic. There are no doubt ways that we could allow more to run in parallel,
+        /// but the current tradeoff is for simplicity of code and "obvious correctness" than something that is subtle, fast, and wrong.</remarks>
+        private readonly object _gate = new object();
+
+        /// <summary>
+        /// The number of active batch scopes. If this is zero, we are not batching, non-zero means we are batching.
+        /// </summary>
+        private int _activeBatchScopes = 0;
+
+        private readonly List<(string path, MetadataReferenceProperties properties)> _metadataReferencesAddedInBatch = new List<(string path, MetadataReferenceProperties properties)>();
+        private readonly List<(string path, MetadataReferenceProperties properties)> _metadataReferencesRemovedInBatch = new List<(string path, MetadataReferenceProperties properties)>();
+        private readonly List<ProjectReference> _projectReferencesAddedInBatch = new List<ProjectReference>();
+        private readonly List<ProjectReference> _projectReferencesRemovedInBatch = new List<ProjectReference>();
+
+        private readonly Dictionary<string, VisualStudioAnalyzer> _analyzerPathsToAnalyzers = new Dictionary<string, VisualStudioAnalyzer>();
+        private readonly List<VisualStudioAnalyzer> _analyzersAddedInBatch = new List<VisualStudioAnalyzer>();
+        private readonly List<VisualStudioAnalyzer> _analyzersRemovedInBatch = new List<VisualStudioAnalyzer>();
+
+        private readonly List<Func<Solution, Solution>> _projectPropertyModificationsInBatch = new List<Func<Solution, Solution>>();
+
+        private string _assemblyName;
+        private string _displayName;
+        private string _filePath;
+        private CompilationOptions _compilationOptions;
+        private ParseOptions _parseOptions;
+        private bool _hasAllInformation = true;
```

Comment:
>private bool _hasAllInformation = true; [](start = 8, length = 39)

Why is the semantics of this bool?

Reply:
When the project system is loading a project and it's not loaded, or if they know that something has failed in a way that means they can't get full information (NuGet restore failed, design time build failed, etc.), they should clear this bit.

The name parallels [this property](https://github.com/dotnet/roslyn/blob/3c69b4bb837c4998028711291f63068c9bbf57f7/src/Workspaces/Core/Portable/Workspace/Solution/ProjectState.cs#L314) which is why it's called this. That API is still internal but once we make it public we'll rename this accordingly to whatever name we end up with.